homework-1/src/task_2.ts

```diff
@@ -0,0 +1,74 @@
+import fs from 'fs';
+import split2 from 'split2';
+import { pipeline, PassThrough, Transform } from 'stream';
+import csv from 'csvtojson';
+
+const PATH_TO_CSV = './nodejs-hw1-ex1.csv';
+const OUTPUT_FILE = './csvOutput.txt';
+
+let header: null | string = null;
+
+const testStream = new PassThrough();
+
+testStream.on('data', (data) => {
+  console.log(data.toString());
+});
+
+const fakeDataBase = (data: string) => {
+  return new Promise<string>((resolve) => {
+    setTimeout(() => {
+      resolve(data);
+    }, 1000);
+  });
+};
+
+const writeToDBStream = new Transform({
+  transform: async (chunk: Buffer, _, callback) => {
+    const dataRow = chunk.toString();
+    try {
+      const response = await fakeDataBase(dataRow);
```

Comment:
I do not like the approach when someone start to doing side effects inside Transform stream. For me it's like a pure function which should have predictable result for the same input. There is one point.
Another point is that you write each chunk sequentially: first you are waiting when db will response after that you pass chunk to the next stream. I suggest, as for me, the more efficient solution:
- Implement custom write stream
- Write chunks to file manually using `writableStream.write()` method (do not forget to check writableStream backpressure)
- You can even use internal buffer to optimise writing to database (to collect some chunks and write them as a bulk, many database support this approach).

There is [a good article](https://nodejs.org/en/docs/guides/backpressuring-in-streams/) to understand what is stream backpressure.

Reply:
Thanks for the explanation. Rewrote with custom implementation of readable stream. I hope I understood correctly what you meant by "use internal buffer"